### ISSUE TITLE
No IP not working

### ISSUE THREAD

**Issue by sterhoke** [qdm12/ddns-updater#44]
Hi,

it seems like NoIp changed something in their http request.

I get badauth response.
When i try with curl it works if i use a base64 encoded password

`curl --header "Authorization: Basic plain_username:base64_password" --header "User-Agent: DDNS-Updater quentin.mcgaw@gmail.com" https://dynupdate.no-ip.com/nic/update?hostname=domain`

`curl --header "Authorization: Basic  base64-encoded-auth-string plain_username:base64_password" --header "User-Agent: DDNS-Updater quentin.mcgaw@gmail.com" https://dynupdate.no-ip.com/nic/update?hostname=domain`
works as well

If i try any other combination it is not working and i get badauth.

I tried setting my password as base64 in the settings file but this will not work either unfortunately.

Regards,
Marcus

**Comment by qdm12** [comment]
It should encode the password as base64, please let me know if it works out for you.

And thanks for reporting with all the debugging details, good job 👍 !

**Comment by sterhoke** [comment]
i had it plain in there.
that's why i tried around.

**Comment by sterhoke** [comment]
just checked again

2020-03-18T01:53:19.530Z	ERROR	invalid response "badauth\r\n"	{"node_id": 0}

I'm getting this error in the log

**Comment by qdm12** [comment]
Haha sorry for messing around with the code formatting :smile:

From their doc it seems that it should be the base 64 encoding of `plain_username:plain_password`.

I changed it so it sends as you described, as basic auth: `plain_username:b64encOf(plain_password)`.

Your config should work but the new Docker image is still baking in Travis (takes like 40 min to build because of an old sqlite C module and cross cpu arch building....), anyway.

If you're impatient you can try with

```sh
docker build -t qmcgaw/ddns-updater https://github.com/qdm12/ddns-updater.git
```

and restart the container

**Comment by sterhoke** [comment]
Build went through on my vm quite fast^^

but i still get the error

2020-03-18T02:10:57.756Z	ERROR	invalid response "badauth\r\n"	{"node_id": 0}

**Comment by qdm12** [comment]
Yeah it's essentially slow to build for other cpu architectures like ARM using an emulator :wink: Building locally for your native architecture is fast enough indeed.

Anyway, I'll try to add `base64-encoded-auth-string` and follow their instructions

**Comment by sterhoke** [comment]
well,

it seems like there is a problem on NoIp's side.
look at this

![image](https://user-images.githubusercontent.com/6491667/76919763-1a2c1300-68ca-11ea-8190-ac2a1a0a917c.png)

I cloned the git and changed the code. Rebuilded but still got the same error.
so i was going back to the roots and checked the curl again....

Also there is a closed issue with noip login with email not working...

It is... :D

**Comment by sterhoke** [comment]
if i change back to username:base64pass it is working as well now 🤣  wtf?

**Comment by sterhoke** [comment]
I changed the delay to 10 seconds.

this resulted in 2 failed authentications.
Afterwards it is working as expected.

so yeah, No-Ip is rejecting my credentials 2 times and then it works. wtf...

**Comment by qdm12** [comment]
Maybe they have a rate limiting? Either way it seems like bad programming on their side as this program really does a simple copy pasta request to noip :smile:

I will still try out to follow their docs base64 encoding the username+password together.

**Comment by sterhoke** [comment]
Just checked again after the night.

the first three requests failed.
the fourth went through....

and there is no other application trying to update the ip, so i dont think this has something to do with a server rate limit.....

imho either their servers are overloaded and not reaching the auth backend or they have other problems right now.
I dont think that this is an issue with the container.... It's NoIp's....

**Comment by qdm12** [comment]
Can you try with curl again? Does it work on first try every time?

If not, can you try using the base64(username:password), see [this](https://www.noip.com/integrate/request)?

Also, maybe 10 seconds is too short, try with 60 seconds? I personally set mine to 5 minutes although my IP address changes once a month usually.

**Comment by sterhoke** [comment]
the last test was with curl

three failed
fouth was ok

tried today again
same behavior.

**Comment by sterhoke** [comment]
> If not, can you try using the base64(username:password), see [this](https://www.noip.com/integrate/request)?

this workes as well after the three failed auths...

I think the only thing you could do to avoid this is sending at least 4 requests,then wait the delay timer and send 4 requests again....

**Comment by qdm12** [comment]
And finally, does it work with the username@password format:

```
http://username:password@dynupdate.no-ip.com/nic/update?hostname=mytest.example.com&myip=192.0.2.25
```

?

**Comment by sterhoke** [comment]
this works as well.

funny enough, it worked on first try without a badauth response.
the curl command failed again with badauth 3 times afterwards....

**Comment by qdm12** [comment]
Alright screw it, I'll just send 4 requests every time :smile:

**Comment by sterhoke** [comment]
I wonder if someone else is experiencing the same behaviour.

Because if this is really an issue on NoIp's side the 4 requests are just a dirty workaround but not really a solution.
Did you made a tiket with NoIp already by any chance? I would create one if not...

Tell me when you are ready with the request loop and i'll test for you... :)

**Comment by qdm12** [comment]
I've changed the format to `https://username:password@dynupdate.no-ip.com/nic/update...` as it wasn't clear as if that format failed as well the first 3 tries. Do you mind trying it (either with your browser or with the latest docker image)? Otherwise I'll setup the 3+1 tries.

**Comment by sterhoke** [comment]
Hey,

it seems like the 3 failed logins are per domain.

i set up a new hostname yesterday and even after successfully updating my main hostname I got 3 failed logins for the new hostname as well.

Domain one:
1st: badauth
2nd: badauth
3rd: badauth
4th: nochange [IP]

Domain two (testet 10 seconds after successfull on domain one):
1st: badauth
2nd: badauth
3rd: badauth
4th: nochange [IP]

i dont think the way you do the http request does not really matter as I'm getting the badauth with all the different versions we tried on the shell.


If i try via browser it works on first try.
with curl only the fourth try will not produce an error.



So in summary:

Browser works everytime
via shell it will produce 3 failed updates


btw.... username:base64_pass works in the browser as well without problems

**Comment by sterhoke** [comment]
´https://username:password@dynupdate.no-ip.com´  first try
´https://username:base64_pass@dynupdate.no-ip.com´ first try

´curl --header "Authorization: Basic  username:password" --header "User-Agent: DDNS-Updater quentin.mcgaw@gmail.com" https://dynupdate.no-ip.com/nic/update?hostname=hostname.net´
three failed

´curl --header "Authorization: Basic  base64-encoded-auth-string username:base64_pass" --header "User-Agent: DDNS-Updater quentin.mcgaw@gmail.com" https://dynupdate.no-ip.com/nic/update?hostname=hostname.net´
three failed

**Comment by qdm12** [comment]
Can you try with curl using the user agent `Mozilla/5.0 (Linux; Android 10; PH-1) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/80.0.3987.149 Mobile Safari/537.36` (my phone browser)? Maybe they blocked the ddns updater user agent for whatever reason.

**Comment by sterhoke** [comment]
same behaviour

**Comment by qdm12** [comment]
And `https://username:password@dynupdate.no-ip.com` using curl doesn't work on first try right?

**Comment by sterhoke** [comment]
correct.

three times badauth. the fourth try is successfull

**Comment by qdm12** [comment]
So working in the end? :smile:

**Comment by sterhoke** [comment]
yep. works now after you changed from header to inline authentication ;)
thank you again
